added main functionality for slots

dem-volodymyr opened the pull request.

dem-volodymyr committed:
added main functionality for slots

gearbeagel commented on backend/Dockerfile:
```diff
@@ -17,6 +17,8 @@ RUN pip install --upgrade pip && pip install -r requirements.txt
 
 RUN chmod +x scripts/wait-for-db.sh
+RUN chmod +x scripts/wait-for-sonar.sh
+RUN chmod +x scripts/init-db.sh
```
Did you add this due to issues on your machine?

dem-volodymyr replied:
Yeah , but unfortunately it didn’t help me.

gearbeagel commented on backend/slots/models.py:
```diff
@@ -1,3 +1,45 @@
+    bet_amount = models.DecimalField(max_digits=10, decimal_places=2)
+    payout = models.DecimalField(max_digits=10, decimal_places=2, default=0.00)
+    result = models.JSONField()  # Store the spin result as JSON
+    win_data = models.JSONField(null=True, blank=True)  # Win information
```
Comments are redundant here.

dem-volodymyr replied:
👍

gearbeagel commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,45 @@
+    class Meta:
+        model = Spin
+        fields = ['id', 'user', 'bet_amount', 'payout', 'result', 'win_data', 'timestamp']
+        read_only_fields = ['id', 'user', 'payout', 'result', 'win_data', 'timestamp']
```
I checked your model. No need to rewrite all of the fields.

```
class SpinSerializer(serializers.ModelSerializer):
    user = UserSerializer(read_only=True)

    class Meta:
        model = Spin
        fields = '__all__'
        read_only_fields = ['id', 'user', 'payout', 'result', 'win_data', 'timestamp']
```

gearbeagel commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,45 @@
+    bet_size = serializers.DecimalField(max_digits=10, decimal_places=2, min_value=0.01)
+
+
+class RegistrationSerializer(serializers.ModelSerializer):
```
We already have registration. Remove this.

dem-volodymyr replied:
sorry I forgot to remove this part during the transfer

gearbeagel commented on backend/slots/models.py:
```diff
@@ -1,3 +1,45 @@
+    @classmethod
+    def create_default_symbols(cls):
+        """Create default symbols that match the frontend."""
+        symbols = [
```
Remove image paths. I use built-in components for generating reels. Also, use IDs instead of names to identify the reels.

gearbeagel commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,45 @@
+from django.contrib.auth.models import User
+
+
+class UserSerializer(serializers.ModelSerializer):
```
This is redundant and already exists.

gearbeagel replied:
Still redundant, already exists. Remove your own implementation of UserSerializer and use an existing one if needed

gearbeagel commented on backend/slots/urls.py:
```diff
@@ -0,0 +1,19 @@
+)
+
+router = DefaultRouter()
+router.register(r'users', UserViewSet, basename='user')
```
Remove this.

gearbeagel commented on backend/slots/services.py:
```diff
@@ -0,0 +1,149 @@
+from django.contrib.auth.models import User
+
+
+class ReelService:
```
Comments are not to be left in the main codebase.

gearbeagel commented on backend/slots/services.py:
Remove comments from the file. They are not to be in the codebase.

gearbeagel commented on backend/slots/views.py:
```diff
@@ -1,3 +1,113 @@
 
-# Create your views here.
+
+class RegistrationView(generics.CreateAPIView):
```
Remove this.

gearbeagel commented on backend/slots/views.py:
```diff
@@ -1,3 +1,113 @@
+        )
+
+
+class UserViewSet(viewsets.ReadOnlyModelViewSet):
```
Remove this.

gearbeagel commented on docker-compose.yml:
Did you make changes to accommodate to your machine?

dem-volodymyr replied:
Yeah

gearbeagel commented on backend/slots/services.py:
```diff
@@ -0,0 +1,149 @@
+        return total_payout
+
+
+class SlotMachineService:
```
In the future, error handling should be made for cases when user's balance is insufficient for the bet.

gearbeagel commented on backend/slots/models.py:
```diff
@@ -1,3 +1,45 @@
+    win_data = models.JSONField(null=True, blank=True)  # Win information
+    timestamp = models.DateTimeField(auto_now_add=True)
+
+    def __str__(self):
```
We kind of do not have a username?
Check [`user/models.py`](https://github.com/gearbeagel/FEPsino/blob/main/backend/user/models.py).

gearbeagel reviewed (changes requested):
It would be appericated if you could fix the issues I have listed. Thanks in advance.

gearbeagel reviewed (commented):
Also, it would be great if you removed everything related to MacOS issues, that wasn't helpful to you.

NikitaPash commented on backend/slots/models.py:
```diff
@@ -1,3 +1,45 @@
 
-# Create your models here.
+
+class Symbol(models.Model):
```
1. Symbol model breaks SRP of SOLID because it handles both creation and storage of symbols
2. Adding new symbols should be possible via the admin panel (you already have it, as far as I can tell, so just make sure adding new symbols via the Admin panel is working)
3. Logic for loading default symbols can and should be separated
To adhere best practices, I suggest next:

```
class Symbol(models.Model):
    name = models.CharField(max_length=50, unique=True)
    payout_multiplier = models.DecimalField(max_digits=5, decimal_places=2, default=1.0)

    def __str__(self):
        return self.name
```
 
 create fixture in slots/fixture/symbols.json with content that looks something like this:
 ```
 [
    {
        "model": "slots.symbol",
        "pk": 1,
        "fields": {
            "name": "star",
            "payout_multiplier": "3.00"
        }
    },
    {
        "model": "slots.symbol",
        "pk": 2,
        "fields": {
            "name": "heart",
            "payout_multiplier": "2.50"
        }
    },
    {
        "model": "slots.symbol",
        "pk": 3,
        "fields": {
            "name": "cherry",
            "payout_multiplier": "2.00"
        }
    },
    {
        "model": "slots.symbol",
        "pk": 4,
        "fields": {
            "name": "gem",
            "payout_multiplier": "1.50"
        }
    },
    {
        "model": "slots.symbol",
        "pk": 5,
        "fields": {
            "name": "citrus",
            "payout_multiplier": "1.00"
        }
    }
]
```

command that should load the defaults:
```
python manage.py loaddata symbols.json
```

NikitaPash commented on backend/slots/models.py:
```diff
@@ -1,3 +1,45 @@
+    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='spins')
+    bet_amount = models.DecimalField(max_digits=10, decimal_places=2)
+    payout = models.DecimalField(max_digits=10, decimal_places=2, default=0.00)
+    result = models.JSONField()  # Store the spin result as JSON
```
I'm not sure we have to save results in a separate field, as bet_amount, payout and win_data can already be consideret a result of the spin

Also clarify what is stored in win_data

NikitaPash commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,45 @@
+class SymbolSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Symbol
+        fields = ['id', 'name', 'image_path']
```
Don't forget to remove image_path here

NikitaPash commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,45 @@
+
+class SpinRequestSerializer(serializers.Serializer):
+    bet_size = serializers.DecimalField(max_digits=10, decimal_places=2, min_value=0.01)
+
```
fix naming for more clarity

class SpinRequestSerializer(serializers.Serializer):
    bet_amount = serializers.DecimalField(max_digits=10, decimal_places=2, min_value=0.01)
    
Also add some validation logic for max bet size and check if user balance is greater than the bet size. Or clarify where the validation is implemented

NikitaPash commented on backend/slots/views.py:
```diff
@@ -1,3 +1,113 @@
+
+class SpinViewSet(viewsets.GenericViewSet):
+    serializer_class = SpinSerializer
+    permission_classes = [IsAuthenticated]
```
add authentication_classes = [JWTAuthentication]

NikitaPash commented on backend/slots/views.py:
```diff
@@ -1,3 +1,113 @@
+        serializer = SpinRequestSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        bet_size = serializer.validated_data['bet_size']
```
follow common naming for bet_amount (or rename every occurence to bet_size)

NikitaPash commented on backend/slots/views.py:
```diff
@@ -1,3 +1,113 @@
+
+class SymbolViewSet(viewsets.ReadOnlyModelViewSet):
+    queryset = Symbol.objects.all()
+    serializer_class = SymbolSerializer
```
add authentication_classes = [JWTAuthentication]

NikitaPash commented on docker-compose.yml:
```diff
@@ -9,7 +9,7 @@ services:
       - "8000:8000"
     command: >
-      sh -c "./scripts/wait-for-db.sh &&
+      chmod +x ./scripts/wait-for-db.sh && sh -c "./scripts/wait-for-db.sh &&
```
we should discuss change to docker-related files at the upcoming weekly meeting

DmBilyk reviewed (changes requested):
FIx Code Readability: Modularize repetitive or complex code blocks for better readability and maintainability.
Fix Error Handling: Add error handling for database operations and external dependencies (e.g., scripts).

MariaHusak commented on backend/slots/services.py:
```diff
@@ -0,0 +1,149 @@
+
+        for row in horizontal:
+            for sym in row:
+                if row.count(sym) > 2:  # Potential win
```
2 is a magic number .It is better to make MIN_SYMBOLS_FOR_WIN = 2 a constant at the class or module level.

MariaHusak commented on backend/slots/services.py:
```diff
@@ -0,0 +1,149 @@
+
+class SlotMachineService:
+    def __init__(self):
+        from .models import Symbol
```
You are re-importing the models. You can move the models to the top . This will make the code look cleaner.

MariaHusak commented on backend/slots/services.py:
```diff
@@ -0,0 +1,149 @@
+        hits = {}
+        horizontal = self.flip_horizontal(result)
+
+        for row in horizontal:
```
I'm not sure, but there might be an error in the loop. If horizontal has identical rows, it will return the index of the first occurrence, not the current one. That is, even if you iterate over the second row, which is the same as the first one — .index(row) will return 0.

MariaHusak reviewed (changes requested):
There are a lot of unnecessary comments in the code, which makes it difficult to read. Please remove the unnecessary comments.

dem-volodymyr committed:
updated changes functionality for slots

dem-volodymyr committed:
updated changes functionality for slots

dem-volodymyr committed:
Deleted unnecessary comments and added unit-tests

dem-volodymyr committed:
Improve Code Readability and Error Handling for models

dem-volodymyr commented:
@DmBilyk 'FIx Code Readability: Modularize repetitive or complex code blocks for better readability and maintainability.
Fix Error Handling: Add error handling for database operations and external dependencies (e.g., scripts).'
Can you check?
https://github.com/gearbeagel/FEPsino/pull/25/commits/233fca441df1c3a449cc110902bdcc570ab83dd9

gearbeagel commented on backend/slots/urls.py:
```diff
@@ -0,0 +1,18 @@
+    path('', include(router.urls)),
+    path('register/', RegistrationView.as_view(), name='register'),
+    path('schema/', SpectacularAPIView.as_view(), name='schema'),
+    path('docs/', SpectacularSwaggerView.as_view(url_name='schema'), name='swagger-ui'),
```
remove swagger-related endpoints, as they are already implemented

dem-volodymyr replied:
fixed swagger-related endpoints on https://github.com/gearbeagel/FEPsino/pull/25/commits/381a22b32dbd92a1b9f80486b472d310a9bf9084

gearbeagel commented on backend/slots/models.py:
```diff
@@ -1,3 +1,21 @@
+
+class Spin(models.Model):
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='spins')
```
You are using built-in user model while our project implements a custom one. Fix by importing correct model:

```
from core.settings import AUTH_USER_MODEL
```
and use this model instead:
```
    user = models.ForeignKey(AUTH_USER_MODEL, on_delete=models.CASCADE, related_name='spins')
```

dem-volodymyr replied:
used the desired user on https://github.com/gearbeagel/FEPsino/commit/381a22b32dbd92a1b9f80486b472d310a9bf9084

gearbeagel commented on backend/slots/admin.py:
```diff
@@ -1,3 +1,13 @@
+
+@admin.register(Symbol)
+class SymbolAdmin(admin.ModelAdmin):
+    list_display = ('name', 'image_path', 'payout_multiplier')
```
remove image_path reference

dem-volodymyr replied:
deleted image path on https://github.com/gearbeagel/FEPsino/commit/381a22b32dbd92a1b9f80486b472d310a9bf9084

gearbeagel commented on backend/slots/urls.py:
add your urls to core.urls:
```
    path('api/slots/', include('slots.urls')),
```

dem-volodymyr replied:
added my urls to core on https://github.com/gearbeagel/FEPsino/commit/381a22b32dbd92a1b9f80486b472d310a9bf9084

gearbeagel commented on backend/slots/urls.py:
```diff
@@ -0,0 +1,18 @@
+from rest_framework.routers import DefaultRouter
+from drf_spectacular.views import SpectacularAPIView, SpectacularSwaggerView
+from .views import (
+    RegistrationView, UserViewSet,
```
remove user-related imports

dem-volodymyr replied:
removed all user imports https://github.com/gearbeagel/FEPsino/commit/381a22b32dbd92a1b9f80486b472d310a9bf9084

gearbeagel commented on backend/slots/urls.py:
```diff
@@ -0,0 +1,18 @@
+
+urlpatterns = [
+    path('', include(router.urls)),
+    path('register/', RegistrationView.as_view(), name='register'),
```
remove user-related urls

dem-volodymyr replied:
deleted user urls on https://github.com/gearbeagel/FEPsino/commit/381a22b32dbd92a1b9f80486b472d310a9bf9084

gearbeagel commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,39 @@
+
+
+class SpinRequestSerializer(serializers.Serializer):
+    bet_amount = serializers.DecimalField(max_digits=10, decimal_places=2, min_value=0.01, max_value=1000.00)
```
use Decimal() class to set min_value and max_value:
```
from decimal import Decimal

...

    bet_amount = serializers.DecimalField(max_digits=10, decimal_places=2, min_value=Decimal(0.01), max_value=Decimal(1000.00))
```

gearbeagel replied:
This issue still persists. Please, solve it.

dem-volodymyr replied:
i resolved it

gearbeagel commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,39 @@
+        read_only_fields = ['id', 'name', 'payout_multiplier']
+
+class SpinSerializer(serializers.ModelSerializer):
+    user = UserSerializer(read_only=True)
```
Use existing user-related functionality

gearbeagel commented on backend/slots/services.py:
```diff
@@ -0,0 +1,288 @@
+        try:
+            symbols = Symbol.objects.all()
+            self.reel_service = ReelService(symbols)
+            self.user = UserManager()
```
remove redundant reference of UserManager

dem-volodymyr replied:
removed this part of the code https://github.com/gearbeagel/FEPsino/commit/381a22b32dbd92a1b9f80486b472d310a9bf9084

gearbeagel replied:
You did not.

gearbeagel commented on backend/slots/services.py:
```diff
@@ -0,0 +1,288 @@
+        """Update user balance for a new bet with error handling."""
+        try:
+            user.balance -= Decimal(bet_amount)
+            user.total_wager += Decimal(bet_amount)
```
rewrite the logic to fit our user-related functionality

gearbeagel commented on backend/slots/services.py:
```diff
@@ -0,0 +1,288 @@
+    def _update_user_balance_for_win(self, user, payout):
+        """Update user balance for a win with error handling."""
+        try:
+            user.balance += payout
```
rewrite the logic to fit our user-related functionality

gearbeagel commented on backend/slots/services.py:
```diff
@@ -0,0 +1,288 @@
+        """Process a single spin of the slot machine."""
+        try:
+            # Validate user balance
+            if user.balance < Decimal(bet_amount):
```
rewrite the logic to fit our user-related functionality

gearbeagel commented on backend/slots/services.py:
```diff
@@ -0,0 +1,288 @@
+                'result': result,
+                'win_data': win_data,
+                'payout': payout,
+                'current_balance': float(user.balance)
```
rewrite the logic to fit our user-related functionality

gearbeagel reviewed (changes requested):
Next issues should be resolved. Also do not resolve comments by yourself. It should be done by the one who wrote them.

Correct workflow should be:
1. You see a comment, stating an issue
2. You fix the issue
3. You reply to the comment or send a reaction. Both should indicate that you either fixed an issue or have some questions regarding it
4. You do not resolve the comment and wait for the reviewer to resolve it or update the issue comment

dem-volodymyr committed:
Refactor user model references and improve error handling.

Updated user model references to use `settings.AUTH_USER_MODEL` for better flexibility and compatibility. Enhanced error handling in the slots service by adding attribute checks and logging warnings. Simplified admin configurations and removed unnecessary fields in serializers and URLs to streamline the application.

gearbeagel committed:
Merge branch 'main' into slots_backend

dem-volodymyr committed:
Updated serializer for custom_user

dem-volodymyr committed:
Merge remote-tracking branch 'origin/slots_backend' into slots_backend

dem-volodymyr committed:
all working now

gearbeagel commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,34 @@
+
+
+class SpinRequestSerializer(serializers.Serializer):
+    bet_amount = serializers.DecimalField(max_digits=10, decimal_places=2, min_value=0.01, max_value=1000.00)
```
```backend  | /usr/local/lib/python3.11/site-packages/rest_framework/fields.py:990: UserWarning: max_value should be a Decimal instance.```
```backend  | /usr/local/lib/python3.11/site-packages/rest_framework/fields.py:992: UserWarning: min_value should be a Decimal instance.```     

Replace this with:


```python
bet_amount = serializers.DecimalField(max_digits=10, decimal_places=2, min_value=Decimal("0.01"), max_value=Decimal("1000.00")
```

gearbeagel commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,34 @@
+
+class SpinSerializer(serializers.ModelSerializer):
+    user_id = serializers.ReadOnlyField(source='user.id')
+    user_email = serializers.ReadOnlyField(source='user.email')
```
I do not comprehend, why would you need these two fields.

dem-volodymyr replied:
ok

gearbeagel commented on backend/slots/serializers.py:
```diff
@@ -0,0 +1,34 @@
+
+    class Meta:
+        model = Spin
+        fields = ['id', 'user', 'user_id', 'user_email', 'payout', 'result', 'win_data', 'timestamp', 'bet_amount']
```
I strongly advise you to pass on the profile instance.

dem-volodymyr replied:
ok

gearbeagel reviewed (changes requested):
Please, address the issues I stated.

dem-volodymyr committed:
deleted unnecessary usermanager() from slotmachineservice class

dem-volodymyr committed:
idk what

MariaHusak reviewed (approved):
The code doesn't look bad and I don't see any major bugs in the code.

NikitaPash reviewed (changes requested):
I'll wait for you to fix Victoria's remarks and then would like to review the code myself

gearbeagel committed:
Merge branch 'main' into slots_backend

DmBilyk reviewed (changes requested):
In general, everything looks good, you just need to fix what Victoria remarked

dem-volodymyr committed:
resolved issues and edited symbols for better code reading.

NikitaPash reviewed (approved):
I still don't get the logic behind result.
```
{
  "success": true,
  "spin_id": "f9247a8c-4f62-408e-a0fe-ce23fb52cbaf",
  "result": {
    "0": [
      0,
      4,
      3
    ],
    "1": [
      2,
      1,
      4
    ],
    "2": [
      2,
      3,
      1
    ],
    "3": [
      2,
      3,
      0
    ],
    "4": [
      0,
      4,
      3
    ]
  },
  "win_data": {
    "1": [
      "cherry",
      [
        1,
        2,
        3
      ]
    ]
  },
  "payout": 600,
  "current_balance": 1660
}
```

by what logic cherry here is under the key "1" in win data if it is under index 3 in symbols endpoint and under index 2 in frontend mapping. Also the fact that win data returns null if no combinations were met, is concerning. Nevertheless, considering deadlines of this project, I will leave the decision to Victoria, because she will be the one to connect this to front end

gearbeagel reviewed (approved):
I'll give this code a shot and try connecting it to the frontend. Will message you, if anything goes wrong.
